src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxEquivalenceTests.cs

```diff
@@ -448,5 +455,112 @@ public void TestExpressionBodiedMethod()
+        [InlineData("", "#nullable enable")]
+        [InlineData("", "#nullable restore")]
+        public void TestNullableDirectives_DifferentDirectives(string firstDirective, string secondDirective)
+        {
+            var tree1 = SyntaxFactory.ParseSyntaxTree($@"
+{firstDirective}
+class C
+{{
+}}");
+            var tree2 = SyntaxFactory.ParseSyntaxTree($@"
+{secondDirective}
+class C
+{{
+}}");
+
+            VerifyNotEquivalent(tree1, tree2, topLevel: true);
+            VerifyNotEquivalent(tree1, tree2, topLevel: false);
+            VerifyEquivalent(tree1.GetRoot(), tree2.GetRoot(), ignoreChildNode: k => k == SyntaxKind.NullableDirectiveTrivia);
+        }
+
+        [Theory, WorkItem(38694, "https://github.com/dotnet/roslyn/issues/38694")]
+        [InlineData("#nullable enable", "#nullable disable")]
+        [InlineData("#nullable enable", "#nullable restore")]
+        [InlineData("#nullable disable", "#nullable restore")]
+        [InlineData("", "#nullable disable")]
+        [InlineData("", "#nullable enable")]
+        [InlineData("", "#nullable restore")]
+        public void TestNullableDirectives_Nested(string firstDirective, string secondDirective)
+        {
+            var tree1 = SyntaxFactory.ParseSyntaxTree($@"
+class C
+{{
+    void M()
+    {{
+{firstDirective}
+    }}
+}}");
+            var tree2 = SyntaxFactory.ParseSyntaxTree($@"
+class C
+{{
+    void M()
+    {{
+{secondDirective}
+    }}
+}}");
+
+            VerifyNotEquivalent(tree1, tree2, topLevel: true);
+            VerifyNotEquivalent(tree1, tree2, topLevel: false);
+            VerifyEquivalent(tree1.GetRoot(), tree2.GetRoot(), ignoreChildNode: k => k == SyntaxKind.NullableDirectiveTrivia);
+        }
+
+        [Theory, WorkItem(38694, "https://github.com/dotnet/roslyn/issues/38694")]
+        [InlineData("#nullable enable")]
+        [InlineData("#nullable disable")]
+        [InlineData("#nullable restore")]
+        public void TestNullableDirectives_TopLevelIdentical(string directive)
+        {
+            var tree1 = SyntaxFactory.ParseSyntaxTree($@"
+class C
+{{
+    void M()
+    {{
+{directive}
+        Console.WriteLine(1234);
+    }}
+}}");
+            var tree2 = SyntaxFactory.ParseSyntaxTree($@"
+class C
+{{
+    void M()
+    {{
+{directive}
+    }}
+}}");
+
+            VerifyEquivalent(tree1, tree2, topLevel: true);
+            VerifyNotEquivalent(tree1, tree2, topLevel: false);
+        }
+
+        [Fact, WorkItem(38694, "https://github.com/dotnet/roslyn/issues/38694")]
+        public void TestNullableDirectives_InvalidDirective()
+        {
+            var tree1 = SyntaxFactory.ParseSyntaxTree(@"
+class C
+{
+    void M()
+    {
+#nullable invalid
+    }
+}");
+            var tree2 = SyntaxFactory.ParseSyntaxTree(@"
+class C
+{
+    void M()
+    {
+    }
+}");
+
+            VerifyNotEquivalent(tree1, tree2, topLevel: true);
+            VerifyNotEquivalent(tree1, tree2, topLevel: false);
+        }
```

Comment:
consider testing:

- a `#nullable enable` that changes to `#nullable enable warnings`
- a `#nullable enable annotations` that changes to `#nullable enable warnings`
- a file with multiple nullable directives, where an additional directive is added or removed